pvlib/solarposition.py

```diff
@@ -1210,3 +1210,50 @@ def hour_angle(times, longitude, equation_of_time):
     )).total_seconds() / 3600. for t in times])
     timezone = times.tz.utcoffset(times).total_seconds() / 3600.
     return 15. * (hours - 12. - timezone) + longitude + equation_of_time / 4.
+
+
+def _hours(times, hour_angle, longitude, equation_of_time):
+    timezone = times.tz.utcoffset(times).total_seconds() / 3600.
+    return (hour_angle - longitude - equation_of_time/4.)/15. + 12. + timezone
+
+
+def _times(times, hours):
+    return np.array([(dt.timedelta(hours=h) + t.tz.localize(
+        dt.datetime(t.year, t.month, t.day)
+    )) for h, t in zip(hours, times)])
```

Comment:
This loop using pure python is likely to be painfully slow and defeat the purpose of an analytical approach.

Reply:
I think I can use numpy arrays with `timedelta64[s]` here instead

Reply:
Nice catch @wholmgren , I get about a 70x speed improvement (from 43.5 ms to 605us) to  by using numpy `datetime64` and `timedelta64[ns]`. I wonder where else we could use these? There is an oddity that `hours` is a `Float64Index` from pandas, not just floats, which bothers me, but ok for now.

See this script that shows improvements with numpy over looping.

```python
In [1]: import pandas as pd

In [2]: import datetime as dt

In [3]: import numpy as np

In [4]: d = pd.DatetimeIndex(start='2013-01-01 00:00:00', end='2013-12-31 23:59:59', freq='H', tz='Etc/GMT+8')

In [5]: h = np.loadtxt('../sunrise_sunset_transit.txt', dtype=np.dtype([('sunrise', float), ('sunset', float), ('transit', float)]))

In [6]: sr = pd.Float64Index(h['sunrise'])

In [7]: %paste
def _times(times, hours):
    return np.array([(dt.timedelta(hours=h) + t.tz.localize(
        dt.datetime(t.year, t.month, t.day)
    )) for h, t in zip(hours, times)])

## -- End pasted text --

In [8]: %timeit _times(d, sr)
43.5 ms ± 354 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [9]: del _times

In [10]: %paste
def _times(times, hours):
    """helper converts hours from an array of floats to localized times"""
    tz_info = times.tz
    return pd.DatetimeIndex(
        times.tz_localize(None).values.astype('datetime64[D]')
        + (hours.values * 3600. * 1.e9).astype('timedelta64[ns]'), tz=tz_info)

## -- End pasted text --

In [11]: %timeit _times(d, sr)
605 µs ± 5.07 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```